net: lwm2m: Fix SenML CBOR object link encoding

broglep-work opened the pull request:
The previous patch could not be applied, so I had to manually re-apply the patch. I kept generated code & modification of it as separate commits for now to ease review, will squash when reviewed. 

further, clang-format has been applied to patch commit (99f2f689a82a72f3327ff8a9051cc4a3f3a8920a) 

fixes #52527

broglep-work committed:
net: lwm2m: fix senml cbor object link encoding

use text format instead of tlv format

Signed-off-by: Pascal Brogle <pascal.brogle@husqvarnagroup.com>

broglep-work committed:
net: lwm2m: regenerate cbor code using zcbor

with the updated cddl for object links

Signed-off-by: Pascal Brogle <pascal.brogle@husqvarnagroup.com>

broglep-work committed:
net: lwm2m: support senml cbor object link

use "vlo" map key

Signed-off-by: Pascal Brogle <pascal.brogle@husqvarnagroup.com>

broglep-work committed:
net: lwm2m: update cbor code patch file

based on generated code

Signed-off-by: Pascal Brogle <pascal.brogle@husqvarnagroup.com>

rlubos commented:
CC @VeijoPesonen

rlubos reviewed (commented):
Thank you @broglep-work for the contribution. The changes look reasonable to me, however I'd like @VeijoPesonen to review the code as he's the original SenML CBOR contributor.

However, it seems that the patch file is not complete? We still hit some compliance issues with the autogenerated files.

VeijoPesonen commented:
@broglep-work Thanks, I'm in the middle of checking the PR but looks really good thus far. In the meanwhile would you please address the whitespace issues compliance check reports.

VeijoPesonen reviewed (commented):
Looks good to me. @broglep-work Greatly appreciated :) I'll approve once you have fixed the minor whitespace issues.

VeijoPesonen commented:
@broglep-work As the compliance check tells please provide a commit body on each separate commit message to describe the change a bit more.

broglep-work commented:
Any suggestions how to fix those compliance check? 

I think it does not make much sense manually modifying patch files to make checkpatch happy, nor it does seam that there is an easy way to ignore that particular file. And trying to reverse engineer & modify the original files so that the patch file makes checkpatch happy feels worthless.

And there is another warning about max line length in the generated code, where clang-format is happy with exceeding 100 chars but checkpatch is not. As this is generated code, I'd also rather ignore those checks, or fix zcbor so that it does not produces lines longer than 100, instead of changing that code manually (which then causes potential future patch applys too fail)

rlubos commented:
I had the impression that the whole idea for the patch file was to summarize the changes needed to align the autogenerated file with Zephyr requirements? @VeijoPesonen How was this handled in the past? @oyvindronningstad Perhaps you could chip in given you've been involved in the process in the past?

> And there is another warning about max line length in the generated code, where clang-format is happy with exceeding 100 chars but checkpatch is not.

Is that really the case for clang-format? There is an explicit limit set in the config file to 100 chars/line:
https://github.com/zephyrproject-rtos/zephyr/blob/main/.clang-format#L30

broglep-work commented:
Not sure how well that patch approach worked in the past, but for my work it was not of much use. A document that described which high level modifications need to be applied to the generated would maybe have been more helpful. (Such as replace `9223372036854775807L` with `INT64_MIN`, Remove hardcoded limit of 99 with config, etc.)

My best idea at the moment is move all that code generation stuff outside of the current directory into a dedicated one and ignore that directory for checkpatch

> Is that really the case for clang-format? There is an explicit limit set in the config file to 100 chars/line:

As far as I know clang-format deviates from that 100 char limit in sensible cases, which in this case it seems to be good judgement of clang-format, re-formatting that few lines to have less than 100 lines makes the already hard to read code (because autogenerated) even more difficult to read.

Ideally that auto-generated file could also be excluded from checkpatch. If not, would probably be okish to modify the file manually so that we don't exceed 100 lines, but this will make the patch file approach of applying changes even more brittle than it already is

broglep-work committed:
net: lwm2m: robuster senml code modification

use comby to transform generated senml code.
this better documents the desired transformations
and is more robust than relying on patch files.

Signed-off-by: Pascal Brogle <pascal.brogle@husqvarnagroup.com>

broglep-work committed:
net: lwm2m: move generated senml code

in order to ignore it for checkpatch. generated code is already formatted
with clang-format and making sure that clang-format never exceeds max line
length is near too impossible and results in sigificant harder to read
code (clang-format has good reasons to sometimes exceed max line length).
manually modifying generated code to not exceed max line length is
worthless.

Signed-off-by: Pascal Brogle <pascal.brogle@husqvarnagroup.com>

broglep-work commented:
After many hours wasted due to checkpatch (and failure to tweak clang-format to respect max line length at all costs) , I think I've come to a solution that is as good I can get it.  

At least we now have much better approach for the code transformation using http://comby.dev which documents the transformations and should be more robust for the future. 

It is also easier to reason about the changes to the generated code. At least while writing the transformation I had several thought that the transformation could be maybe be better (thought that did not cross my mind I was when working with the patch file and reverse-engineer the intended transformations from the patch file). In some places where `ZCBOR_ARRAY_SIZE()` is used, we could probably just use `CONFIG_LWM2M_RW_SENML_CBOR_RECORDS`

rlubos commented on subsys/net/lib/lwm2m/CMakeLists.txt:
```diff
@@ -71,8 +71,8 @@ zephyr_library_sources_ifdef(CONFIG_LWM2M_RW_CBOR_SUPPORT
     lwm2m_rw_senml_cbor.c
-    lwm2m_senml_cbor_decode.c
-    lwm2m_senml_cbor_encode.c
+        generated/lwm2m_senml_cbor_decode.c
+        generated/lwm2m_senml_cbor_encode.c
```
I don't think an extra indent is needed here.

broglep-work replied:
missed that one, have fixed it (was introduced refactoring / auto-format)

rlubos commented:
Adding more reviewers into the loop given the PR now references new util for autogenerated ode tranformation. I'm ok with this, but wanted to hear from others.

VeijoPesonen commented:
Placing the generated files to a separate folder and making Git to ignore it sounds reasonable for me. In my opinion introducing one more external tool like comby is no go but that's not for me to decide. Looks like a handy tool though.

SeppoTakalo commented:
I'm not familiar with this tool `comby`

Is it really so much better that we should convert this specific converter to use it while nobody else in Zephyr do not.
It does not even appear in common repositories for common Linux distros.

Why not try to do the same transformation with a bit familiar tool that already has use in Zephyr: https://docs.zephyrproject.org/latest/develop/tools/coccinelle.html

broglep-work commented:
As comby is not directly used as part of the regular code workflow, I don't see much of an issue. We already depend on zcbor python package as an external tool. 

A nice solution would be to integrate the code generation and transformation into cmake, in that case I agree, we should more carefully consider which tools are in use (and make sure that the get properly installed as part of the regular developer setup)

As long as we have a workflow to run a script occasionally, generate some code, that we then review & commit, I think we should be able to use tools that get the job done quickly. Comby seemed to be a popular choice, although I spent maybe 15 minutes of research. There might be other tools out there that are more in line with the technologies/ecosystem of zephyr

broglep-work commented:
> Why not try to do the same transformation with a bit familiar tool that already has use in Zephyr: https://docs.zephyrproject.org/latest/develop/tools/coccinelle.html

sounds reasonable, anyone here that has experience with that tool that can try convert from comby to coccinelle?

carlescufi commented:
@broglep-work thanks very much for the proposal. I have added the TSC label so that we can discuss the approach there. If you'd like to join today, here are the details:
https://github.com/zephyrproject-rtos/zephyr/wiki/Technical-Steering-Committee-%28TSC%29

SeppoTakalo reviewed (dismissed):
After a second thought, I'm proposing that we accept this patch as it is.

Yes, I would like to see the formatting done with Coccinelle, but we should open a issue for it, and fix it later. 
This PR fixes a real issue in code and it is causing usability issues at least with AVSystem.

We can convert the Comby script to Coccinelle later.

galak commented:
I think if the files are committed into the tree they should conform to the coding style.

broglep-work commented:
> I think if the files are committed into the tree they should conform to the coding style.

I generally agree, the underlying issue here however is we have conflicting style guidelines, I opened #52712 for this. So the only way to resolve this at the moment is break one of the guidelines.

SeppoTakalo commented:
> I think if the files are committed into the tree they should conform to the coding style.

Maybe this rule should be understood that files, we expect humans to touch.

After all, we are now talking about generated parser which we should not allow anyone to directly edit.

carlescufi commented:
@broglep-work during last week's TSC there seemed to be clear agreement as to not allow yet another (i.e. comby) source code parser/reformatter. I suggest the following as alternatives:

- Commit the untouched files, then immediately commit the modified (by hand or coccinelle) files so we keep track of the diff with Git
- Commit the modified files and a coccinelle script
- Commit the modified files and a `.patch`file that diffs the generatd vs modified files

Any of those should work, but please remove the .comby stuff.

carlescufi commented on .checkpatch.conf:
```diff
@@ -29,3 +29,4 @@
 --ignore MULTISTATEMENT_MACRO_USE_DO_WHILE
 --ignore ENOSYS
 --ignore IS_ENABLED_CONFIG
+--exclude subsys/net/lib/lwm2m/generated
```
remove this

carlescufi commented on subsys/net/lib/lwm2m/codegen/lwm2m_senml_cbor_comby.conf:
```diff
@@ -0,0 +1,120 @@
+#
```
remove this file

carlescufi reviewed (changes requested):
see my comment with proposed solutions

broglep-work commented:
> @broglep-work during last week's TSC there seemed to be clear agreement as to not allow yet another (i.e. comby) source code parser/reformatter. I suggest the following as alternatives:
> 
> * Commit the untouched files, then immediately commit the modified (by hand or coccinelle) files so we keep track of the diff with Git
> * Commit the modified files and a coccinelle script
> * Commit the modified files and a `.patch`file that diffs the generatd vs modified files
> 
> Any of those should work, but please remove the .comby stuff.

I will have to revisit this at a later point in time as I had to move on to other topics in the meantime. 

Based on what I've learned, probably only option 1) will be possible, after maybe performing some tool assisted modification and after that, tweak the code manually to pass code styles (as there is no guarantee that clang-format produces correct style, and the expectation is that the developer after that manually adjustes the code, see #52712) 
While this works, it is not really nice developer experience, and the next developer that needs touch that code needs to bite the bullet.

Option 2) might have the issue that coccinelle might not be able to perform formatting changes ( e.g break lines to adhere to coding style). So a single coccinelle transformation might not be sufficient and some additional tool is required (or manually changes). But as I mentioned, I'm not familiar with that tool, maybe it is also capable of transforming the code to adhere to style guides with some human help where clang-format itself fails.
Option 3) Using patch files has the issue that it can trigger checkpatch to think it is c code and complain about it (which happened in this PR). Same risk probably also exists with Option 2). It also is not really good in encoding the intended modification and not robust enough that the patch can be applied for modified / regenerated code. So also not of much use. 

The only option that seems valuable to me and worth investing time is removing the generated code from tree and generate it on-the-fly at compile time. This will resolve probably all problems with the current solution.

carlescufi commented:
> The only option that seems valuable to me and worth investing time is removing the generated code from tree and generate it on-the-fly at compile time. This will resolve probably all problems with the current solution.

I understand but at the moment there is no consensus regarding code generation at build/generation time in Zephyr as a project, especially in cases like this, so this will require some discussion and may take a while.

rlubos commented:
@broglep-work Would you mind if we took your commits and open a separate PR, w/o introducing new tools (comby) to get the bugfix merged? We'd like to move things forward.

SeppoTakalo commented:
I have assigned this PR to our internal backlog, so we will refactor this PR into a separate PR without introducing dependency to a new tool.
After all, the bugfix is valid and the CDDL code is already provided so the hard work is already done.

broglep-work commented:
I am back to work after the holiday break, and found this a nice christmas / new year gift, thanks for moving this forward in my absence

jheiskan81 commented:
@broglep-work Thanks for good contribute which was easy to push forward.
